pkg/remoteimageasync/puller.go

```diff
@@ -0,0 +1,70 @@
+package remoteimageasync
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/warm-metal/container-image-csi-driver/pkg/metrics"
+	"k8s.io/klog/v2"
+)
+
+// sessionChan and completedChan both closed here
+func RunPullerLoop(
+	ctx context.Context,
+	sessionChan chan *PullSession,
+	completedChan chan string,
+) {
+	go func() {
+		<-ctx.Done()
+		close(sessionChan) // only close this once
+	}()
+	go func() {
+		wg := sync.WaitGroup{}
+		for {
+			ses, ok := <-sessionChan // ctx not observed for shut down, this sleep breaks when sessionChan is closed
+			if !ok {                 // sessionChan closed, shut down loop
+				wg.Wait()            // wait for running goroutines (which observe cancellation)
+				close(completedChan) // only close this once, after completions submitted (wg.Wait() returns)
```

Comment:
no, sessionChan must be closed by the sender. `RunPullerLoop()` is the receiving activity. this part of Go channels design is a mess IMHO. it should allow checking for closed when writing just like when reading by returning second optional result. instead, it panics. the theory is that a producer should not close a channel without coordinating with other producers (complexity). i would prefer if that coordination was as i just described instead of needing more synchronization primitives and one-off design.

Reply:
writing to the channel already has to synchronize so it should offer the "is closed" check for free IMHO. we could still implement our own coordination if we could not start an activity unless we knew we could submit that to the channel... but now we need to implement that for all cases. the example here was one where it was a bit complex to determine how to close the channels because there could always be another `NodePublishVolume()` call to come it after `sessionChan` was closed.